### PR TITLE
ath79: add support for Senao Engenius EnStationAC v1

### PR DESCRIPTION
**Pull request by mcprat**
FCC ID: A8J-ENSTAC

Engenius EnStationAC is an outdoor wireless access point/bridge with 2 gigabit ethernet ports,
5 GHz only wireless, external ethernet switch, internal antenna plates and proprietery PoE.

Specification:

  - QCA9557
  - 40 MHz reference clock
  - 16 MB FLASH			MX25L12845EMI-10G
  - 2x 64 MB RAM		NT5TU32M16FG
  - UART at J10			(unpopulated)
  - 2x GbE ports with POE	(Atheros AR8033 switch)
  - 5 GHz, 2x2, 26dBm		(QCA9882 PCI card)
  - internal antenna plates	(19 dbi, directional)
  - 7 LEDs, 1 button		(power, eth, wlan, RSSI) (reset)

MAC addresses:

  eth0	*:d3   art 0x0/0x6
  eth1	*:d4   ???
  wlan	*:d5   ???

  The device label lists first ETH MAC and WLAN 5 GHz MAC

  Even though only 1 MAC representing the vendor is found on flash
  The OEM software reports these MACs for the rest of the ifconfig.

Installation:

  2 ways to flash factory.bin from OEM:

  - Connect ethernet directly to board (the non POE port)
      this is LAN for all images
  - if you get Failsafe Mode from failed flash:
      only use it to flash Original firmware from Engenius
      or risk kernel loop or halt which requires serial cable

  Method 1: Firmware upgrade page:

    OEM webpage at 192.168.1.1
    username and password "admin"
    Navigate to "Firmware" page from left pane
    Click Browse and select the factory.bin image
    Upload and verify checksum
    Click Continue to confirm and wait 3 minutes

  Method 2: Serial to load Failsafe webpage:

    After connecting to serial console and rebooting...
    Interrupt uboot with any key pressed rapidly
    execute `run failsafe_boot` OR `bootm 0x9fd70000`
    wait a minute
    connect to ethernet and navigate to
    "192.168.1.1/index.htm"
    Select the factory.bin image and upload
    wait about 3 minutes

Return to OEM:

  If you have a serial cable, see Serial Failsafe instructions

  *DISCLAIMER*
  The Failsafe image is unique to Engenius boards.
  If the failsafe image is missing or damaged this will not work
  DO NOT downgrade to ar71xx this way, can cause kernel loop or halt

  The easiest way to return to the OEM software is the Failsafe image
  If you dont have a serial cable, you can ssh into openwrt and run

  `mtd -r erase fakeroot`

  Wait 3 minutes
  connect to ethernet and navigate to 192.168.1.1/index.htm
  select OEM firmware image from Engenius and click upgrade

Format of OEM firmware image:

  The OEM software of EnStationAC is a heavily modified version
  of Openwrt Altitude Adjustment 12.09. One of the many modifications
  is to the sysupgrade program. Image verification is performed
  simply by the successful ungzip and untar of the supplied file
  and name check and header verification of the resulting contents.
  To form a factory.bin that is accepted by OEM Openwrt build,
  the kernel and rootfs must have specific names...

    openwrt-ar71xx-enstationac-uImage-lzma.bin
    openwrt-ar71xx-enstationac-root.squashfs

  and begin with the respective headers (uImage, squashfs).
  Then the files must be tarballed and gzipped.
  The resulting binary is actually a tar.gz file in disguise.
  This can be verified by using binwalk on the OEM firmware images,
  ungzipping then untaring.

  Newer EnGenius software requires more checks but their script
  includes a way to skip them, otherwise the tar must include
  a text file with the version and md5sums in a deprecated format.

  The OEM upgrade script is at /etc/fwupgrade.sh.

  OKLI kernel loader is required because the OEM software
  expects the kernel to be no greater than 1536k
  and the factory.bin upgrade procedure would otherwise
  overwrite part of the kernel when writing rootfs.

Note on PLL-data cells:

  The default PLL register values will not work
  because of the external AR8033 switch between
  the SOC and the ethernet PHY chips.

  For QCA955x series, the PLL registers for eth0 and eth1
  can be see in the DTSI as 0x28 and 0x48 respectively.
  Therefore the PLL registers can be read from uboot
  for each link speed after attempting tftpboot
  or another network action using that link speed
  with `md 0x18050028 1` and `md 0x18050048 1`.

  For eth0 at 1000 speed, the value returned was
  ae000000 but that didn't work, so following
  the logical pattern from the rest of the values,
  the guessed value of a3000000 works better.

Tested from master, all link speeds functional

Signed-off-by: Michael Pratt <mpratt51@gmail.com>

### CODE REVIEW

**Comment by mpratt14** [comment]
@adschm the 3rd board, last one for now

some notes:

ethernet leds are controlled by the external switch

ath10k / wifi node has wrong MAC address (not the one on label)

uboot has wrong pll for eth0 speed 1000

openwrt/LuCI reports that cpu is a qca9558 but it is qca9557

adding a touched file to factory.bin recipe, failsafe.bin
which allows overriding of the extra checks that this board has
Here is the OEM upgrade script
https://gist.github.com/mpratt14/b810163865432891c5d9b34bf81a1ab8
at line 184 is the bypass for the checks
the other 2 boards have smaller scripts with less checks

For ath10k I picked the package `ath10k-firmware-qca988x-ct-full-htt`
is this the best one? any reason similar boards don't use this one?


boot log:
https://gist.github.com/mpratt14/c958ff804f92161eda84948ac0c78c17

File: target/linux/ath79/generic/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -158,6 +158,13 @@ engenius,ens202ext-v1)
 	ucidef_set_led_rssi "rssimedium" "RSSIMEDIUM" "$boardname:amber:rssimedium" "wlan0" "33" "100"
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "$boardname:green:rssihigh" "wlan0" "67" "100"
 	;;
+engenius,enstationac-v1)
+	ucidef_set_led_wlan "wlan" "WLAN" "$boardname:green:wlan" "phy0tpt"
```
Comment (adschm) [inline target/linux/ath79/generic/base-files/etc/board.d/01_leds:162]: You have a trigger in device tree, so this line can just be dropped.

File: target/linux/ath79/image/generic.mk [hunk 2]
```diff
@@ -860,6 +861,18 @@ define Device/engenius_ens202ext-v1
 endef
 TARGET_DEVICES += engenius_ens202ext-v1
 
+define Device/engenius_enstationac-v1
+  $(Device/engenius_loader_okli)
+  SOC := qca9557
+  DEVICE_MODEL := EnStationAC
+  DEVICE_VARIANT := v1
+  DEVICE_PACKAGES := ath10k-firmware-qca988x-ct-full-htt rssileds
```
Comment (adschm) [inline target/linux/ath79/image/generic.mk:869]: Please use ath10k-firmware-qca988x-ct like for the other devices. If you want to use a different package, you will have to pull it via opkg or build custom images.

File: target/linux/ath79/generic/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (mpratt14) [reply target/linux/ath79/generic/base-files/etc/board.d/01_leds]: this makes it the same as ens202ext so I merged them

File: target/linux/ath79/image/generic.mk (hunk 2 above)
Comment (mpratt14) [reply target/linux/ath79/image/generic.mk]: just wondering why one is the default over the other?

Comment (adschm) [reply target/linux/ath79/image/generic.mk]: Well, one has to be default. I assume the smallest package was taken ...

**Comment by mpratt14** [comment]
@adschm This is ready BTW

> art partition wifi data has wrong MAC address (not the one on label)

Sorry if this confused you
it said "wifi / ath10k node has wrong MAC address" I meant it "had" the wrong mac address which is why I have it in the case with the mac patch in caldata script

**Comment by mpratt14** [comment]
On second thought, theres no particular reason that the POE port has to be WAN, so when you merge just remove all changes to 02_network script, the default is fine

**Comment by adschm** [comment]
> On second thought, there's no particular reason that the POE port has to be WAN, so when you merge just remove all changes to 02_network script, the default is fine

The default case in 02_network will just produce the inverted assignment for LAN and WAN.

**Comment by mpratt14** [comment]
@adschm 
Rebased, removed changes to 02_network, disabled mdio1. Should be complete now.

**Comment by mpratt14** [comment]
@adschm removed unnecessary mdio1 node, fixed wifi node name, rebased

File: target/linux/ath79/dts/qca9557_engenius_enstationac-v1.dts [hunk 3]
```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca955x.dtsi"
```
Comment (adschm) [inline target/linux/ath79/dts/qca9557_engenius_enstationac-v1.dts:7]: Move this include before others.

File: target/linux/ath79/dts/qca9557_engenius_enstationac-v1.dts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_power: power {
+			label = "enstationac-v1:amber:power";
```
Comment (adschm) [inline target/linux/ath79/dts/qca9557_engenius_enstationac-v1.dts:36]: drop model: prefix for led labels.

File: target/linux/ath79/image/generic.mk [hunk 5]
```diff
@@ -56,6 +56,7 @@ endef
 define Build/engenius-tar-gz
 	-[ -f "$@" ] && \
 	mkdir -p $@.tmp && \
+	touch $@.tmp/failsafe.bin && \
```
Comment (adschm) [inline target/linux/ath79/image/generic.mk:59]: Since this affects several devices, it should go into a separate commit with proper description.

**Comment by mpratt14** [comment]
@adschm split into 2 commits, updated and rebased

File: target/linux/ath79/image/generic.mk [hunk 6]
```diff
@@ -891,6 +891,18 @@ define Device/engenius_ens202ext-v1
 endef
 TARGET_DEVICES += engenius_ens202ext-v1
 
+define Device/engenius_enstationac-v1
+  $(Device/engenius_loader_okli)
+  SOC := qca9557
+  DEVICE_MODEL := EnStationAC
+  DEVICE_VARIANT := v1
+  DEVICE_PACKAGES := ath10k-firmware-qca988x-ct rssileds
```
Comment (adschm) [inline target/linux/ath79/image/generic.mk:900]: You are missing kmod-ath10k-ct or -smallbuffers variant here, adding just the firmwaer.

Comment (mpratt14) [reply target/linux/ath79/image/generic.mk:900]: hmm...I see that it selects kmod-ath10k automatically as a dependency...

why is kmod-ath10k-ct not set as dependency for ath10k-firmware-qca988x-ct?

Comment (adschm) [reply target/linux/ath79/image/generic.mk:900]: > hmm...I see that it selects kmod-ath10k automatically as a dependency...

Where?

> why is kmod-ath10k-ct not set as dependency for ath10k-firmware-qca988x-ct?

E.g. because you might want kmod-ath10k-ct-smallbuffers instead.

Comment (mpratt14) [reply target/linux/ath79/image/generic.mk:900]: Sorry I'm confusing with my custom build

kmod-ath10k is selected automatically when I have ath10k-firmware-qca988x (not ct version) I think...

**Comment by mpratt14** [comment]
This unit is currently being used, I need to switch it out with something else so I can redo the pll-data values, now that I know more about it

**Comment by mpratt14** [comment]
fixed firmware packages list

corrected eth0 pll register according to what I discovered in my other PRs

updated commit message

rebased

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Comment by mpratt14** [comment]
@ynezz same here, if Bjorn's mtdsplit patches will not be accepted in time to revert

```
				compatible = "openwrt,uimage", "denx,uimage";
				openwrt,ih-magic = <0x4f4b4c49>;
```

back to "openwrt,okli"
